casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala

```diff
@@ -68,56 +68,120 @@ object BlockAPI {
+        ListeningNameContinuationResponse(status = "Success",
+                                          blockResults = blocksWithActiveName,
+                                          length = blocksWithActiveName.length.toLong)
+
+    implicit val channelCodec: Codec[Channel] = serializeChannel.toCodec
+    MultiParentCasperConstructor.withCasper[F, ListeningNameContinuationResponse](
+      casperResponse(_, channelCodec),
+      ListeningNameContinuationResponse(status = "Error: Casper instance not available"))
+  }
+
+  private def getMainChainFromTip[F[_]: Monad: MultiParentCasper: Log: SafetyOracle: BlockStore]
+    : F[IndexedSeq[BlockMessage]] =
+    for {
+      estimates   <- MultiParentCasper[F].estimator
+      tip         = estimates.head
+      internalMap <- BlockStore[F].asMap()
+      mainChain   = ProtoUtil.getMainChain(internalMap, tip, IndexedSeq.empty[BlockMessage])
+    } yield mainChain
+
+  private def getDataWithBlockInfo[F[_]: Monad: MultiParentCasper: Log: SafetyOracle: BlockStore](
+      runtimeManager: RuntimeManager,
+      sortedListeningName: Channel,
+      block: BlockMessage)(implicit channelCodec: Codec[Channel]): F[Option[DataWithBlockInfo]] =
+    if (isListeningNameReduced(block, immutable.Seq(sortedListeningName))) {
+      val stateHash =
+        ProtoUtil.tuplespace(block).get
+      val data =
+        runtimeManager.getData(stateHash, sortedListeningName)
+      for {
+        blockInfo <- getBlockInfoWithoutTuplespace[F](block)
+      } yield Option[DataWithBlockInfo](DataWithBlockInfo(data, Some(blockInfo)))
+    } else {
+      none[DataWithBlockInfo].pure[F]
+    }
+
+  private def getContinuationsWithBlockInfo[
+      F[_]: Monad: MultiParentCasper: Log: SafetyOracle: BlockStore](
+      runtimeManager: RuntimeManager,
+      sortedListeningNames: immutable.Seq[Channel],
+      block: BlockMessage)(
+      implicit channelCodec: Codec[Channel]): F[Option[ContinuationsWithBlockInfo]] =
+    if (isListeningNameReduced(block, sortedListeningNames)) {
+      val stateHash =
+        ProtoUtil.tuplespace(block).get
+      val continuations: Seq[(Seq[BindPattern], Par)] =
+        runtimeManager.getContinuation(stateHash, sortedListeningNames)
+      val continuationInfos = continuations.map(continuation =>
+        WaitingContinuationInfo(continuation._1, Some(continuation._2)))
+      for {
+        blockInfo <- getBlockInfoWithoutTuplespace[F](block)
+      } yield
+        Option[ContinuationsWithBlockInfo](
+          ContinuationsWithBlockInfo(continuationInfos, Some(blockInfo)))
+    } else {
+      none[ContinuationsWithBlockInfo].pure[F]
+    }
+
+  private def isListeningNameReduced(
+      block: BlockMessage,
+      sortedListeningName: immutable.Seq[Channel])(implicit channelCodec: Codec[Channel]) = {
+    val serializedLog =
+      block.body.fold(Seq.empty[Event])(_.commReductions)
+    val log =
+      serializedLog.map(EventConverter.toRspaceEvent).toList
+    log.exists {
```

Comment:
Is this enough? For example, if either the produce or consume is already in the tuplespace when the other half arrives to make a comm event then we still want to report that right? So do we need to look for the `COMM` case as well?

Reply:
Great catch; will do